### DISCUSSION TITLE
Is it possible to use `ws` and `http` DownstreamScheme at the same time?

### DISCUSSION THREAD

**Discussion by slaterx** [ThreeMammals/Ocelot#2153] [Q&A] [answered]
## Expected Behavior / New Feature

I'm wondering if it's possible to have the same API endpoint supporting `ws` and `http` at the same time.

Example:
```json
    {
      "DownstreamPathTemplate": "/api/foo/{everything}",
      "DownstreamScheme": [ "http", "ws" ],
      "DownstreamHostAndPorts": [
        {
          "Host": "foo",
          "Port": 19999
        }
      ],
      "UpstreamPathTemplate": "/api/foo/{everything}",
      "UpstreamHttpMethod": [ "Put", "Delete", "Get", "Post" ]
    }
```

**Comment by raman-m** [comment, marked as answer]
Hello Gleidson,
I'm curious about the questions you're asking us. What is your ultimate objective? Is there a reason you cannot deploy two instances on different ports?

**Firstly**, the WebSocket protocol is indeed based on HTTP.
**Secondly**, it's not advisable to mix multiple protocols associated with multiple ports of the same service instance.
**Thirdly**, Ocelot cannot transform WebSocket traffic to HTTP and vice versa due to its pipeline design restrictions, which only allow re-routing WebSocket traffic. However, in theory, a WebSocket to HTTP fallback is possible, but it would require specific settings.

Ocelot is currently capable of performing transformations for both HTTP/HTTPS schemes and WS/WSS schemes. However, there may be some limitations or specific configurations needed for WebSocket (WS/WSS) scheme transformations.

Ocelot is unable to perform the following tasks:
- Convert WebSocket (`ws`) to HTTP protocol and vice versa, as this feature has not been implemented.
- Convert WebSocket Secure (`wss`) to HTTPS traffic.
- Convert any protocol to another protocol, as such functionality would require substantial development efforts.

Please let us know which gateways are able to convert protocols?
